### DISCUSSION TITLE
Race condition - fetch and update overwriting each other. Need ideas

### DISCUSSION THREAD

**Discussion by bdelamotte** [prisma/prisma#10709] [Q&A] [answered]
We have an API call that needs to fetch a database entry, get a column, and append to that column. The datatype is JSON so we are adding an object to a JSON array and saving it back to the that row. 

We do that in two steps: prisma.findUnique and prisma.update. The problem occurs where the API is called back to back and findUnique grabs the same data twice and then only one value is saved and the other seems to get overwritten. We thought wrapping this in a transaction would fix the issue, but it still happens. Throwing a random sleep in the API call so that the prisma calls are spread out by a few milliseconds fixes the issue but feels like a hack. Essentially this is what seems to happen:

API call 1 - Gets JSON data - currently length 1
API call 2 - Gets JSON data - currently length 1
API call 1 - Saves JSON data - currently length 2 
API call 2 - Overwrites JSON data - currently length 2

Expected behavior in a transaction
API call 1 - Gets JSON data - currently length 1
API call 1 - Saves JSON data - currently length 2 
API call 2 - Gets JSON data - currently length 2
API call 2 - Save JSON data - currently length 3

Any ideas to solve this without using a setTimer() to space the API calls out and thus the prisma calls, would be greatly appreciated.

**Comment by thrant** [comment]
(Major edit after re-reading the question, misunderstood it at first so that the API call sends the data to frontend, someone modifies it, and then sends it back - but this all happens within a single request)

I can think of following solutions:

1) The obvious solution would be to transform the JSON data into a proper SQL structure, allowing you to create new rows without fear of any owerwrites. But I understand that due to various (usually non-technical) reasons this might be not practical or possible. However if you really are getting API calls that are milliseconds apart I guess you're getting a lot of those, and it might be worthwhile to do the refactoring. If say an external party needs the data in JSON, it should be pretty trivial to re-create after pulling the data from the DB.

2) Try writing a custom DB query instead of using Prismas' findUnique & update, for example in MySQL JSON_ARRAY_APPEND() might be what you need

3) This is more of a theoretical solution, but if you must operate on a single (and likely huge) JSON file with a lot of incoming changes and want guarantee that there are no overwrites, you could do a semaphore solution where each request is assigned an ID, and there is another DB column that stores the current "owner" ID. Only the request that "owns" the JSON at any time proceeds to go and read it, modify it, save it, and then releases the semaphore and then the next request grabs the ownership, starts the read, modifies, etc.. but if this is practical or any less work than the rewrite, I dunno. :)

**Comment by bdelamotte** [comment]
Thank you @thrant for taking the time to respond to this. Much appreciated!

Adding some more context.

> The obvious solution would be to transform the JSON data into a proper SQL structure, allowing you to create new rows without fear of any owerwrites. 
>
We went down this path but seemed like storing JSON  in database was best because it is all image metadata and we wanted to keep the order of images contained. We originally had an Images model that would relate back to our other model but there didn't seem to be a nice and efficient way to store order in the database as date modified wouldn't work for us. If the parent model had 20-30 photos and required to be ordered, we thought storing this as a JSON field on the parent model would be the best design. Maybe not though? Perhaps I'm missing a design pattern that could be better and move away from JSON.

> Try writing a custom DB query instead of using Prismas' findUnique & update, for example in MySQL JSON_ARRAY_APPEND() might be what you need
> 
Definitely think we will try this next!

> This is more of a theoretical solution, but if you must operate on a single (and likely huge) JSON file with a lot of incoming changes and want guarantee that there are no overwrites, you could do a semaphore solution where each request is assigned an ID, and there is another DB column that stores the current "owner" ID. Only the request that "owns" the JSON at any time proceeds to go and read it, modify it, save it, and then releases the semaphore and then the next request grabs the ownership, starts the read, modifies, etc.. but if this is practical or any less work than the rewrite, I dunno. :)
> 
I thought about a [similar idea ]( https://blog.revolve.team/2020/09/08/implement-mutex-with-dynamodb/)but by using dynamodb. Perhaps keeping the semaphore / mutex would be best in the same database though. Definitely will think on this.

**Comment by thrant** [comment]
I think that on the short run using MySQL's direct JSON manipulation will be best, but on the long run I would recommend just refactoring the JSON into proper tables.

> If the parent model had 20-30 photos and required to be ordered, we thought storing this as a JSON field on the parent model would be the best design. Maybe not though? Perhaps I'm missing a design pattern that could be better and move away from JSON.

First of all, I think Postgres can store arrays natively. But I guess you (like me) are using MySQL, and there I've had some success with simply using an order number int field. The usecase is one where I to pull a list of stuff to a frontend UI, manipulate it there (add / delete / reorder), and then save it. Just before saving on the backend I set the orderNumber to be a running number from 1 up for the full array of items, in the order which the array came in. Then when the list is needed again, I just pull it out and use orderBy orderNumber to get the same order out.

Alternatively every modification could've been sent immediately to the backend, necessary order changes done to the item in question and all subsequent items (see https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#atomic-number-operations for easy incrementing and decrementing), and ideally the changes propagated to everyone who's currently reading the same list.

If there's a lot of concurrent edits these solutions too might need some mutex-style stuff, but that starts going back to the business logic of whose changes take precedent, should conflicts be bounced back or silently merged somehow, etc, since it doesn't make sense to offer a manipulation of a list which someone else just re-orders right away.. there some user-specific views or sorts might make more sense.

> Perhaps keeping the semaphore / mutex would be best in the same database though. Definitely will think on this.

Oh yeah, definitely keep it in the same DB if you go that route.

**Comment by bdelamotte** [comment, marked as answer]
I am hoping this helps someone. Here is the solution we wrote that solved the issue for us in Postgres. We appended to the photos JSON column using the `||` operator in Postgres like so:
```
await prisma.$executeRaw`update "Listing" set photos = photos ||  ${JSON.stringify(imgMetadata)}::jsonb
          where id = ${listingId}`
```

**Reply by bdelamotte** [reply]
Big thanks to @thrant for pointing me in the right direction.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
